streamer/src/nonblocking/quic.rs

```diff
@@ -519,10 +558,72 @@ async fn setup_connection(
     }
 }
 
+async fn packet_batch_sender(
+    packet_sender: Sender<PacketBatch>,
+    packet_receiver: AsyncReceiver<PacketAccumulator>,
+    exit: Arc<AtomicBool>,
+    stats: Arc<StreamStats>,
+) {
+    trace!("enter packet_batch_sender");
+    loop {
+        let mut packet_batch = PacketBatch::with_capacity(PACKET_BATCH_SIZE);
+
+        stats
+            .total_packets_allocated
+            .fetch_add(PACKET_BATCH_SIZE, Ordering::Relaxed);
+
+        let mut last_sent = Instant::now();
+        loop {
+            if exit.load(Ordering::Relaxed) {
+                return;
+            }
+            let elapsed = last_sent.elapsed();
+            if packet_batch.len() >= PACKET_BATCH_SIZE
+                || (!packet_batch.is_empty() && elapsed.as_millis() >= COALESCE_MS)
+            {
+                let len = packet_batch.len();
+                if let Err(e) = packet_sender.send(packet_batch) {
+                    stats
+                        .total_packet_batch_send_err
+                        .fetch_add(1, Ordering::Relaxed);
+                    trace!("Send error: {}", e);
+                } else {
+                    stats
+                        .total_packet_batches_sent
+                        .fetch_add(1, Ordering::Relaxed);
+                    trace!("Sent {} packet batch", len);
+                }
+                break;
+            }
+
+            let res = packet_receiver.try_recv();
+            if let Ok(packet_accumulator) = res {
+                unsafe {
+                    packet_batch.set_len(packet_batch.len() + 1);
+                }
+
+                let i = packet_batch.len() - 1;
+                *packet_batch[i].meta_mut() = packet_accumulator.meta;
+                for packet_bytes in packet_accumulator.bytes {
+                    packet_batch[i].buffer_mut()
+                        [packet_bytes.offset as usize..packet_bytes.end_of_chunk]
+                        .copy_from_slice(&packet_bytes.bytes);
+                }
+
+                if packet_batch.len() == 1 {
+                    last_sent = Instant::now();
+                }
+            } else {
+                sleep(Duration::from_micros(250)).await;
```

Comment:
This is another latency added, can we eliminate it?

Reply:
Unfortunately not, I think. This adds a small amount of latency if we happen to sleep just before the 5ms time limit is up, but even then, only 250 microseconds (testing has failed to find any harm from this, and the sleep time is a small percentage of the 5ms time limit itself), and it is essential to have this short sleep to prevent useless spinning in periods when there are no packets incoming, and also to avoid this async task never yielding.

Reply:
We can't use recv_timeout in the crossbeam channel, because that blocks the worker thread, which we generally don't want to do in async code, and in this case actually causes reliability problems. I have however tried doing an async timeout with the async channel:

`timeout(receiver.recv()).await`

but this turns out to have significantly worse perf.

Reply:
Can you give the sth like the following a try to use fuse and select!?

use futures::{FutureExt, select};

async fn recv_timeout<T>(receiver: &mut Receiver<T>, timeout: Duration) -> Option<T> {
    let mut recv_fut = receiver.recv().fuse();
    let mut timeout_fut = futures_timer::Delay::new(timeout).fuse();

    select! {
        val = recv_fut => Some(val),
        _ = timeout_fut => None,
    }
}

Reply:
Sure, let me try that.

Reply:
Ok, just realized my original timeout vs sleep comparison was not 1-1 and in any case, the thin-client is pretty variable. Retested with the timeout and found no difference in perf.